internal/venice-client-common/src/main/java/com/linkedin/venice/utils/concurrent/CloseableThreadLocal.java

```diff
@@ -0,0 +1,95 @@
+package com.linkedin.venice.utils.concurrent;
+
+import java.util.Set;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * A wrapper of {@link ThreadLocal} for {@link AutoCloseable} objects.
+ * Java's {@link ThreadLocal} only dereferences the thread-local objects
+ * when the {@link Thread} exits which triggers their garbage collection.
+ * It does not {@code close} the {@link AutoCloseable} thread-local objects.
+ * @param <T> The {@link AutoCloseable} type whose objects will be held by an object of this class.
+ */
+public class CloseableThreadLocal<T extends AutoCloseable> implements AutoCloseable {
+  private static final Logger LOGGER = LogManager.getLogger(CloseableThreadLocal.class);
+  private final Set<T> itemSet = VeniceConcurrentHashMap.newKeySet();
```

Comment:
Why not just use `ConcurrentSkipListSet` instead?

Reply:
I wasn't aware of this. However, we used `ConcurrentSet` from Netty previously - which was deprecated and the recommendation there is to use [ConcurrentHashMap.newKeySet](https://netty.io/4.1/api/io/netty/util/internal/ConcurrentSet.html). We have adopted this in our codebase in `Utils#newConcurrentSet`. I couldn't use the same util class in this PR since it is present in `venice-common` whereas this class is in `venice-client-common`.

Moving Utils class to `venice-client-common` would involve lots of changes and I want to avoid them in this PR

Let me know what you think?

Reply:
So `VeniceConcurrentHashMap` doesn't provide an overrided implementation of `newKeySet`, so `newKeySet` will return a set based on `ConcurrentHashMap` and if we want to leverage the optimization implemented in `VeniceConcurrentHashMap`, we need to re-impl `newKeySet` to use `VeniceConcurrentHashMap` instead of the default `ConcurrentHashMap`:
```
public static <K> KeySetView<K,Boolean> newKeySet() {
        return new KeySetView<K,Boolean>
            (new ConcurrentHashMap<K,Boolean>(), Boolean.TRUE);
    }
``` 

`VeniceConcurrentHashMap` is in `venice-client-common`, and I guess we could use it directly.

Reply:
Currently, I am already using `VeniceConcurrentHashMap`.

The [bug](https://bugs.openjdk.org/browse/JDK-8161372) that `VeniceConcurrentHashMap` solves is only in `computeIfAbsent` function. Do we need to override the `newKeySet` function for that? This would also trigger some class movement since VeniceConcurrentHashMap would need move to `java.util.concurrent` package to access the Package-private constructor for KeySetView
```java
        KeySetView(ConcurrentHashMap<K,V> map, V value) {  // non-public
            super(map);
            this.value = value;
        }
```

Reply:
That is not possible. Only JDK classes are allowed in that package.

Reply:
[This](http://hg.openjdk.java.net/jdk9/jdk9/jdk/rev/fe0d3813e6c3) was the commit that fixed [JDK-8161372](https://bugs.openjdk.org/browse/JDK-8161372). There was a change in `putVal` which is also used in `KeySetView#add`. So, I guess it would have been best to use `ConcurrentSkipListSet` like Gaojie initially suggested. 

However, I need to change the datastructure to a map now to address [this review comment](https://github.com/linkedin/venice/pull/89#discussion_r1012364353). I'll try `ConcurrentSkipListSet` if I decide to remove the unused functions instead

Reply:
SG